### ISSUE TITLE
testdiff3D

### ISSUE THREAD

**Issue by robertjharrison** [m-a-d-n-e-s-s/madness#133]
It's a mix of a TBB generated issue and unsafe programming in the old
model --- basically the current implementation on TBB changed our task
execution model.

The programming model in MADNESS is officially that a task should
never block ... dependencies/communication should be encapsulated in
futures passed as arguments to tasks that can run to completion once
all of the arguments are ready.

However, the old implementation was quite forgiving.  A waiting thread
worked on another task so U could actually specify tasks in any order,
have a task block internally and still be sure that eventually a
dependency would be satisfied (unless U ran out of stack first ...
this is the unsafe aspect).

Using TBB a task/thread cannot work while waiting on a MADNESS future.

I have not delved into the differentiation code enough to see what is
going on, but minimally it needs refactoring so that instead of
blocking while forcing a future it submits another task.  Here is the
traceback
#0  0x00007ffff408dacd in nanosleep ()

```
at ../sysdeps/unix/syscall-template.S:81
```
#1  0x00007ffff40bf034 in usleep (useconds=<optimized out>)

```
at ../sysdeps/unix/sysv/linux/usleep.c:32
```
#2  0x0000000000436c06 in madness::Futurestd::pair<madness::Key<3ul,

madness::GenTe\
nsor<double> > >::get (this=0x7ffff17de710)
 at ../../../src/madness/world/worldthread.h:1084
#3  0x000000000042e2c2 in madness::Derivative<double, 3ul>::do_diff2b

(
    this=0x7ffff17de710, f=0x0, df=0x0, key=..., left=..., center=...
    right=...) at ../../../src/madness/mra/derivative.h:356

   <=============== this needs refactoring to avoid .get() of a
non-ready future.
#4  TBB stuff above here

On Thu, Feb 12, 2015 at 11:24 AM, Nichols A. Romero naromero@gmail.com wrote:

> The hung queue problem is on BG with 1 MPI task, which is based on an old
> MPICH (Jeff will remember which version).
> 
> On Thursday, February 12, 2015, Jeff Hammond jeff.science@gmail.com wrote:
> 
> > I am having the issue on Mac, so it's not a BG thing.
> > 
> > What MPI are you using?  If OpenMPI, I am inclined to wonder if we are
> > abusing MPI and that MPICH notices.
> > 
> > Jeff
> > 
> > On Thu, Feb 12, 2015 at 7:51 AM, Robert Harrison rjharrison@gmail.com
> > wrote:
> > 
> > > moldft runs OK for me on linux with multiple MPI processes regardless
> > > of the optimization level
> > > 
> > > it could be the same bug being expressed in a different way on bg but
> > > it could also be something else
> > > 
> > > On Thu, Feb 12, 2015 at 10:41 AM, Nichols A. Romero naromero@gmail.com
> > > wrote:
> > > 
> > > > I _think_ this is the same bug I see with the nacl input file. If I go
> > > > all
> > > > the way back to Nov. 14, it seems to work. However, when I went back as
> > > > far
> > > > as Dec. 28, the problem was still there.
> > > > 
> > > > On Thursday, February 12, 2015, Robert Harrison rjharrison@gmail.com
> > > > wrote:
> > > > 
> > > > > There does seem to be a problem that has crept in.
> > > > > 
> > > > > Under linux with the intel 14.x compiler using default (-O3 etc)
> > > > > optimization, testdiff3D.mpi hangs reproducibly.
> > > > > 
> > > > > With -O1 it runs.
> > > > > 
> > > > > So either the compiler optimization is exposing a bug of ours or the
> > > > > compiler is injecting its own.
> > > > > 
> > > > > I will trace this down.
> > > > > ##

**Comment by justusc** [comment]
> Using TBB a task/thread cannot work while waiting on a MADNESS future.

This is no longer true. I fixed the issue a few weeks ago.

**Comment by robertjharrison** [comment]
that is where it is still hanging.



On Sun, May 24, 2015 at 3:10 PM, Justus Calvin <notifications@github.com>
wrote:

> Using TBB a task/thread cannot work while waiting on a MADNESS future.
>
> This is no longer true. I fixed the issue a few weeks ago.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/m-a-d-n-e-s-s/madness/issues/133#issuecomment-105049108>
> .
>



-- 
Robert J. Harrison
tel:  865-272-9262

**Comment by justusc** [comment]
I understand that it is hanging there, but MADNESS can definitely run tasks while waiting. I attached to the hung process on my laptop and I see the following:
```
* thread 1: tid = 0x1f31d, 0x00007fff8bd7e5c2 libsystem_kernel.dylib`swtch_pri + 10, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame 0: 0x00007fff8bd7e5c2 libsystem_kernel.dylib`swtch_pri + 10
    frame 1: 0x00007fff85de67df libsystem_pthread.dylib`sched_yield + 11
    frame 2: 0x00000001030bfb17 libtbb_debug.dylib`tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::receive_or_steal_task(long&) + 1735
    frame 3: 0x00000001030beda6 libtbb_debug.dylib`tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::local_wait_for_all(tbb::task&, tbb::task*) + 5206
    frame 4: 0x00000001030bd8f9 libtbb_debug.dylib`tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::wait_for_all(tbb::task&, tbb::task*) + 41
    frame 5: 0x0000000100024cce testdiff2D.mpi`tbb::task::wait_for_all(this=0x00000001049e3b40) + 62 at task.h:744
    frame 6: 0x0000000100024a64 testdiff2D.mpi`madness::ThreadPool::run_task() + 164 at thread.h:1332
    frame 7: 0x00000001000247d5 testdiff2D.mpi`void madness::ThreadPool::await<madness::BindNullaryConstMemFun<madness::FutureImpl<std::__1::pair<madness::Key<2ul>, madness::GenTensor<double> > >, bool> >(probe=0x00007fff5fbf75c0, dowork=true) + 101 at thread.h:1384
    frame 8: 0x0000000100024655 testdiff2D.mpi`void madness::World::await<madness::BindNullaryConstMemFun<madness::FutureImpl<std::__1::pair<madness::Key<2ul>, madness::GenTensor<double> > >, bool> >(probe=0x00007fff5fbf75c0, dowork=true) + 37 at world.h:502
    frame 9: 0x00000001000245e4 testdiff2D.mpi`madness::FutureImpl<std::__1::pair<madness::Key<2ul>, madness::GenTensor<double> > >::get(this=0x000000010382dc40) + 244 at future.h:336
    frame 10: 0x0000000100023d1a testdiff2D.mpi`madness::Future<std::__1::pair<madness::Key<2ul>, madness::GenTensor<double> > >::get(this=0x00007fff5fbf8bf8) + 266 at future.h:596
    frame 11: 0x00000001000213bd testdiff2D.mpi`madness::Derivative<double, 2ul>::do_diff2b(this=0x00007fff5fbfe648, f=0x0000000104156e00, df=0x000000010433be00, key=0x00000001049c3d88, left=0x00000001049c3da8, center=0x00000001049c3e60, right=0x00000001049c3f18) const + 2797 at derivative.h:351
...
```
You can see from frames 2-4 that the thread is trying to steal work form other threads. Hence running tasks in the background. The other thread is doing something similar.

I am not sure why/how it is deadlocking, but I did verify that increasing the number of threads allows the test to run to completion. Generally, it is safe, and even recommended, for tasks to wait on other tasks with TBB since they can work while waiting.

**Comment by robertjharrison** [comment]
Somehow the running of tasks is not functioning sufficiently to ensure
execution in the way the old thread library did.

The traceback I had earlier indicated one thread was hanging while forcing
a future (it was trying to execute a task while waiting but somehow not
making progress).



On Sun, May 24, 2015 at 10:13 PM, Justus Calvin <notifications@github.com>
wrote:

> I understand that it is hanging there, but MADNESS can definitely run
> tasks while waiting. I attached to the hung process on my laptop and I see
> the following:
>
> * thread 1: tid = 0x1f31d, 0x00007fff8bd7e5c2 libsystem_kernel.dylib`swtch_pri + 10, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
>   * frame 0: 0x00007fff8bd7e5c2 libsystem_kernel.dylib`swtch_pri + 10
>     frame 1: 0x00007fff85de67df libsystem_pthread.dylib`sched_yield + 11
>     frame 2: 0x00000001030bfb17 libtbb_debug.dylib`tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::receive_or_steal_task(long&) + 1735
>     frame 3: 0x00000001030beda6 libtbb_debug.dylib`tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::local_wait_for_all(tbb::task&, tbb::task*) + 5206
>     frame 4: 0x00000001030bd8f9 libtbb_debug.dylib`tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::wait_for_all(tbb::task&, tbb::task*) + 41
>     frame 5: 0x0000000100024cce testdiff2D.mpi`tbb::task::wait_for_all(this=0x00000001049e3b40) + 62 at task.h:744
>     frame 6: 0x0000000100024a64 testdiff2D.mpi`madness::ThreadPool::run_task() + 164 at thread.h:1332
>     frame 7: 0x00000001000247d5 testdiff2D.mpi`void madness::ThreadPool::await<madness::BindNullaryConstMemFun<madness::FutureImpl<std::__1::pair<madness::Key<2ul>, madness::GenTensor<double> > >, bool> >(probe=0x00007fff5fbf75c0, dowork=true) + 101 at thread.h:1384
>     frame 8: 0x0000000100024655 testdiff2D.mpi`void madness::World::await<madness::BindNullaryConstMemFun<madness::FutureImpl<std::__1::pair<madness::Key<2ul>, madness::GenTensor<double> > >, bool> >(probe=0x00007fff5fbf75c0, dowork=true) + 37 at world.h:502
>     frame 9: 0x00000001000245e4 testdiff2D.mpi`madness::FutureImpl<std::__1::pair<madness::Key<2ul>, madness::GenTensor<double> > >::get(this=0x000000010382dc40) + 244 at future.h:336
>     frame 10: 0x0000000100023d1a testdiff2D.mpi`madness::Future<std::__1::pair<madness::Key<2ul>, madness::GenTensor<double> > >::get(this=0x00007fff5fbf8bf8) + 266 at future.h:596
>     frame 11: 0x00000001000213bd testdiff2D.mpi`madness::Derivative<double, 2ul>::do_diff2b(this=0x00007fff5fbfe648, f=0x0000000104156e00, df=0x000000010433be00, key=0x00000001049c3d88, left=0x00000001049c3da8, center=0x00000001049c3e60, right=0x00000001049c3f18) const + 2797 at derivative.h:351
> ...
>
> You can see from frames 2-4 that the thread is trying to steal work form
> other threads. Hence running tasks in the background. The other thread is
> doing something similar.
>
> I am not sure why/how it is deadlocking, but I did verify that increasing
> the number of threads allows the test to run to completion. Generally, it
> is safe, and even recommended, for tasks to wait on other tasks with TBB
> since they can work while waiting.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/m-a-d-n-e-s-s/madness/issues/133#issuecomment-105092658>
> .
>



-- 
Robert J. Harrison
tel:  865-272-9262
